### ISSUE TITLE
ethtool can't set options for coalesce or ring sizes

### ISSUE THREAD

**Issue by KyleSanderson** [NixOS/nixpkgs#255759]
### Describe the bug
There's no way to set these critical networking features on a NIC using nix.

### Steps To Reproduce
1. Look for coalesce.
2. Find nothing.

### Expected behavior
The ability to set -G and -K options in a configuration file.

### Additional context
These options are critical for networking performance.

### Notify maintainers
@Ericson2314 @infinisil @roberth 

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.52, NixOS, 23.05 (Stoat), 23.05.3427.e5f018cf150e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.5`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by infinisil** [comment]
Make sure to ping the right people. I'm not a maintainer of this package, and the other people you pinged don't seem to be either.

**Comment by KyleSanderson** [comment]
> Make sure to ping the right people. I'm not a maintainer of this package, and the other people you pinged don't seem to be either.

Sorry - still learning this environment. I pulled your name from https://github.com/NixOS/nixpkgs/blob/e78fd721038da570cf55483c3aa32356a70b6257/maintainers/team-list.nix#L652 .

**Comment by KyleSanderson** [comment]
@bjornfor

**Comment by randomizedcoder** [comment]
@KyleSanderson 

TLDR Here's a systemd oneshot unit that will configure the NIC for you.

```
{ pkgs, ... }:
{
    systemd.services.ethtool-enp3s0f0 = {
    description = "ethtool-enp3s0f0";
    serviceConfig = {
        Type = "oneshot";
        User = "root";
        ExecStart = "${pkgs.ethtool}/bin/ethtool --set-ring enp3s0f0 rx 4096 tx 4096";
    };
    wantedBy = [ "multi-user.target" ];
    };
}

```

For others viewing this post, the reason this is an important feature is that most NICs default to small ring buffers.
```
[das@hp1:~/nixos/hp1]$ ethtool --show-ring enp3s0f0
Ring parameters for enp3s0f0:
Pre-set maximums:
RX:			4096
RX Mini:		n/a
RX Jumbo:		n/a
TX:			4096
TX push buff len:	n/a
Current hardware settings:
RX:			256                    <---- :( sad
RX Mini:		n/a
RX Jumbo:		n/a
TX:			256
RX Buf Len:		n/a
CQE Size:		n/a
TX Push:		off
RX Push:		off
TX push buff len:	n/a
TCP data split:		n/a
```

To improve performance, you want to do this:
```
[das@hp1:~/nixos/hp1]$ sudo ethtool --set-ring enp3s0f0 rx 4096 tx 4096
[sudo] password for das: 

[das@hp1:~/nixos/hp1]$ ethtool --show-ring enp3s0f0
Ring parameters for enp3s0f0:
Pre-set maximums:
RX:			4096            <---- woot woot!
RX Mini:		n/a
RX Jumbo:		n/a
TX:			4096
TX push buff len:	n/a
Current hardware settings:
RX:			4096            <---- woot woot!
RX Mini:		n/a
RX Jumbo:		n/a
TX:			4096
RX Buf Len:		n/a
CQE Size:		n/a
TX Push:		off
RX Push:		off
TX push buff len:	n/a
TCP data split:		n/a
```

A solution on Ubuntu is to do something like: /etc/systemd/system/ethtool.enp3s0f0.service
```
[Unit]
Description=ethtool_tweak
Before=network-pre.target

[Service]
ExecStart=/usr/sbin/ethtool --set-ring enp3s0f0 rx 4096 tx 4096
RemainAfterExit=true
Type=oneshot

[Install]
WantedBy=network.target

```

Therefore, I just tried to do the same thing with NixOS and it works:
```
{ pkgs, ... }:
{
    systemd.services.ethtool-enp3s0f0 = {
    description = "ethtool-enp3s0f0";
    serviceConfig = {
        Type = "oneshot";
        User = "root";
        ExecStart = "${pkgs.ethtool}/bin/ethtool --set-ring enp3s0f0 rx 4096 tx 4096";
    };
    wantedBy = [ "multi-user.target" ];
    };
}
```

The crappy thing about this is that you need to create one of these per NIC.  A nicer solution would be to <somehow> have a nix module that could do this, and it could take a list of NICs.

Hopefully this helps

**Comment by randomizedcoder** [comment]
Actually, this is a little better, because usually these ethtool commands bounce the interface, so you want to run ethtool before the NIC really comes up.
```
{ pkgs, ... }:
{
    systemd.services.ethtool-enp3s0f1 = {
    description = "ethtool-enp3s0f1";
    serviceConfig = {
        Type = "oneshot";
        User = "root";
        ExecStart = "${pkgs.ethtool}/bin/ethtool --set-ring enp3s0f1 rx 4096 tx 4096";
    };
    # wantedBy = [ "multi-user.target" ];
    # https://systemd.io/NETWORK_ONLINE/
    wantedBy = [ "network-pre.target" ];
    };
}
```
